### PR TITLE
ensure that an IW has an associated kernel when retrieved

### PR DESCRIPTION
**Pull request by amunger**
For #

-   [ x ] Pull request represents a single change (i.e. not fixing disparate/unrelated things in a single PR).
-   [ x ] Title summarizes what is changing.
-   [ x ] Has a [news entry](https://github.com/Microsoft/vscode-jupyter/tree/main/news) file (remember to thank yourself!).
-   [ x ] Appropriate comments and documentation strings in the code.
-   [ ] Has sufficient logging.
-   [ ] Has telemetry for enhancements.
-   [ ] Unit tests & system/integration tests are added/updated.
-   [ ] [Test plan](https://github.com/Microsoft/vscode-jupyter/blob/main/.github/test_plan.md) is updated as appropriate.
-   [ ] [`package-lock.json`](https://github.com/Microsoft/vscode-jupyter/blob/main/package-lock.json) has been regenerated by running `npm install` (if dependencies have changed).

### CODE REVIEW

File: src/client/datascience/interactive-window/interactiveWindowProvider.ts [hunk 1]
```diff
@@ -95,6 +95,9 @@ export class InteractiveWindowProvider implements IInteractiveWindowProvider, IA
             result = this.create(resource, mode);
         }
 
+        // make sure the existing IW has a kernel in case it failed to start previously
+        result.ensureKernelReadyPromise();
+
         await result.readyPromise;
```
Comment (rchiodo) [inline src/client/datascience/interactive-window/interactiveWindowProvider.ts:101]: Could you put the 'ensureKernelReadyPromise' call into the getter for 'readyPromise'? Then this code wouldn't have to change.

File: src/client/datascience/interactive-window/interactiveWindow.ts [hunk 2]
```diff
@@ -167,12 +167,28 @@ export class InteractiveWindow implements IInteractiveWindowLoadable {
             this.internalDisposables
         );
         this.internalDisposables.push(kernel);
-        await kernel.start();
+
```
Comment (rchiodo) [inline src/client/datascience/interactive-window/interactiveWindow.ts:170]: This function is rather confusing. Internal state is assigned to its result but it also updates that internal state in itself. 

What if you did this instead (and not assign the _kernelReadyPromise as the result of this function)

```
this._kernelReadyPromise = () => {
    try {
        await kernel.start();
        this.fileInKernel = undefined;
        await this.runIntialization(kernel, this.owner);
    } catch {
        // Do nothing
    }
};
```

Comment (rchiodo) [reply src/client/datascience/interactive-window/interactiveWindow.ts:170]: The exception could be thrown too if we want to make sure it comes out. Maybe like this instead:

```
let exception: Error | undefined;
const readyPromise = () => {
    try {
        await kernel.start();
        this.fileInKernel = undefined;
        await this.runIntialization(kernel, this.owner);
    } catch (e) {
        throw e;
        this._kernelReadyPromise = undefined
    }
};
this._kernelReadyPromise = readyPromise;
await readyPromise;
```

Comment (rchiodo) [reply src/client/datascience/interactive-window/interactiveWindow.ts:170]: Or maybe a simpler way, turn your 'ensureKernelReadyPromise' into a wrapper around 'createKernelReadyPromise' and have it behave like so:
```
public ensureKernelReadyPromise() {
    if (!this._kernelReadyPromise) {
        this._kernelReadyPromise = this.createKernelReadyPromise().catch(() => this._kernelReadyPromise = undefined);
    }
}

Comment (rchiodo) [reply src/client/datascience/interactive-window/interactiveWindow.ts:170]: And then replace all calls to 'createKernelReadyPromise' with the 'ensureKernelReadyPromise'. The error will still be thrown if somebody waits on the _kernelReadyPromise, but it won't prevent the next call from recreating it.

File: src/client/datascience/types.ts [hunk 3]
```diff
@@ -433,6 +433,7 @@ export interface IInteractiveWindow extends IInteractiveBase {
     readonly readyPromise: Promise<void>;
     readonly kernelPromise: Promise<IKernel | undefined>;
     closed: Event<void>;
+    ensureKernelReadyPromise(): Promise<IKernel>;
```
Comment (rchiodo) [inline src/client/datascience/types.ts:436]: I don't think this is necessary as you can do this in the 'readyPromise' getter.

**Review by rchiodo — CHANGES_REQUESTED** [review]
What do you think of my suggestion?

File: src/client/datascience/interactive-window/interactiveWindowProvider.ts (hunk 1 above)
Comment (amunger) [reply src/client/datascience/interactive-window/interactiveWindowProvider.ts:101]: that makes sense. I didn't realize readyPromise was only called from the provider, that makes it easy to test

File: src/client/datascience/interactive-window/interactiveWindow.ts [hunk 4]
```diff
@@ -243,7 +251,7 @@ export class InteractiveWindow implements IInteractiveWindowLoadable {
                 this._controllerReadyPromise.resolve(e.controller);
 
                 // Recreate the kernel ready promise now that we have a new controller
-                this._kernelReadyPromise = this.createKernelReadyPromise();
+                this.ensureKernelReadyPromise();
             },
```
Comment (rchiodo) [inline src/client/datascience/interactive-window/interactiveWindow.ts:255]: This won't 'recreate' it if it's already set. You probably need to clear it here first.

**Review by rchiodo — APPROVED** [review]

**Review by IanMatthewHuff — APPROVED** [review]

**Pull request merged**